vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go

```diff
@@ -63,37 +71,63 @@ type podReplicaCreator struct {
 
 // CanEvict checks if pod can be safely evicted
 func (e *podsEvictionRestrictionImpl) CanEvict(pod *apiv1.Pod) bool {
-	cr, present := e.podsCreators[getPodID(pod)]
+	cr, present := e.podToReplicaCreatorMap[getPodID(pod)]
 	if present {
-		return e.evictionBudget[cr] > 0
+		singleGroupStats, present := e.creatorToSingleGroupStatsMap[cr]
+		if pod.Status.Phase == apiv1.PodPending {
+			return true
+		}
+		if present {
+			shouldBeAlive := singleGroupStats.configured - singleGroupStats.evictionTollerance
+			if singleGroupStats.running-singleGroupStats.evicted > shouldBeAlive {
+				return true
+			}
+			// If all pods are running and eviction tollerance is small evict 1 pod.
+			if singleGroupStats.running == singleGroupStats.configured &&
+				singleGroupStats.evictionTollerance == 0 &&
+				singleGroupStats.evicted == 0 {
+				return true
+			}
+		}
 	}
 	return false
 }
 
 // Evict sends eviction instruction to api client. Retrurns error if pod cannot be evicted or if client returned error
 // Does not check if pod was actually evicted after eviction grace period.
 func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod) error {
-	cr, present := e.podsCreators[getPodID(podToEvict)]
+	cr, present := e.podToReplicaCreatorMap[getPodID(podToEvict)]
 	if !present {
 		return fmt.Errorf("pod not suitable for eviction %v : not in replicated pods map", podToEvict.Name)
 	}
-	if e.evictionBudget[cr] < 1 {
-		return fmt.Errorf("cannot evict pod %v : eviction budget exceeded", podToEvict.Name)
-	}
 
-	eviction := &policyv1.Eviction{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: podToEvict.Namespace,
-			Name:      podToEvict.Name,
-		},
-	}
-	err := e.client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
-	if err != nil {
-		glog.Errorf("failed to evict pod %s, error: %v", podToEvict.Name, err)
-		return err
+	if e.CanEvict(podToEvict) {
```

Comment:
You can rewrite this as 
if !e.CanEvict(podToEvict) {
   retrun fmt.Error...
}
rest of the code